W.I.P. Read compressed file

srbcheema1 opened the pull request:
This PR aims to solve the issue #107

srbcheema1 commented:
This code was written before merging the PR #111 so it is having conflicts. that is not an issue i will have a look at it. This is just to show demo of work. I am working on it. 
This works fine while reading files with `.bz2 compression` using `-i filename` flag or `< filename` .
It is having problem with `cat filename |`
just have a look at it and review if I am going in right direction. And suggest me the changes I can do.
This is only a demo of code. may contain some debug comments. I will reactify it all.

srbcheema1 commented on src/vcf/validator.cpp:
```diff
@@ -193,26 +208,38 @@ namespace ebi
+        in.push(boost::iostreams::bzip2_decompressor());
+        in.push(input);
+        boost::iostreams::copy(in, strr);
+        input.rdbuf(strr.rdbuf());
```
cpp references says that its pointless to make a copy of a stream. so what i am doing here is making a new stream of decompressed input and then giving its buffer to original input stream .
but as you told that files may be large , 
this approach needs to be dropped. 
I will figure think something to work in junks or line by line.

jmmut replied:
Are you sure it's copying the whole buffer? I thought that at first, but then I saw that `rdbuf` just sets a pointer to the buffer, which in my understanding can be empty or not read yet. I wouldn't discard this approach yet

srbcheema1 replied:
I think line 241 `boost::iostreams::copy(in, strr);` will be having issue while copying large file. I will research over it.

srbcheema1 replied:
@jmmut just I am asking for curiosity and knoweledge. what may be the maximum size of file we will be dealing with. I just want to know the extreme limit.

jmmut replied:
A VCF is basically a matrix, one dimension is region in the genome, and the other dimension is the samples (individuals) each cell in the matrix is "what variant does a given sample have in a given genomic position". There's a limit on how many variants you can put, because any species DNA is finite, but in theory you can put an arbitrary number of samples in the same file.

Some big files we have at EVA may be well around 30BG, if I remember correctly, but there's no limit. Some other projects are handling VCFs of hundreds of GB or even a few TB, because they join 10^4 to 10^5 samples in the same file.

We had a bug some time ago that we detected validating a file of 3 billion lines. It was most likely an integer overflow (past 2147 million), just like youtube with gangnam style.

jmmut replied:
that's why it's so important the memory usage. We put some effort to be sure that the memory usage in the validator is constant, so it doesn't increase with time. That way, the file can be arbitrarily big.

srbcheema1 replied:
@jmmut I wanted to discuss something regarding a issue mentioned in gsoc projects. The issue regarding "users can validate their local files only" .
I sent a mail regarding my ideas that can be implemented as far as I understood the problem. Please have a look at them and tell me if I am going in right direction.

srbcheema1 replied:
as the file may be large. I was thinking it to break it into chunks before transferring the file over network. Detailed idea is in the link attached in mail. 
Also I was facing a problem regarding some communication. As I am not able to discuss ideas and doubts properly due to lack of a communication channel. I wanted to know if there is no issue in getting my doubts cleared over mail ?

jmmut replied:
Yes, there's no problem with talking through email, I'll answer you there.

srbcheema1 replied:
@jmmut  there is some issue while reading compressed file.
the `readline` function seems to read multiple lines instead of single line (as it may not get '\n' in compressed format there after the first line) in case of a compressed file.
I have tested it on several test files. it seems that it reads whole of the file in one go.
let us assume it reads whole of the file in one go if it is a compressed file.
So, it is not good behaviour of `readline` if it does so. so we should need to modify` readline` function a bit (if possible).

srbcheema1 commented:
Till now vcf-validator is capable of reading .bz2 compressed files only if provided with -i command line argument. It is unable to read file if provided through stdin. work is in progress. have a look at it.

jmmut commented on src/vcf/validator.cpp:
```diff
@@ -222,17 +249,18 @@ namespace ebi
 
     bool validate(const std::vector<char> &firstLine,
-                  std::istream &input,
+                  boost::iostreams::filtering_istream &input,
```
I'd say filtering `filtering_istream` extends `std::istream`, so we can keep asking for the interface here.

jmmut commented on inc/util/stream_utils.hpp:
```diff
@@ -63,6 +64,24 @@ namespace ebi
     }
 
+    template <typename Container>
+    Container &read_line(boost::iostreams::filtering_istream & stream, Container & container)
```
actually, I think this function is not needed at all, as filtering_istream implements std::istream, right?

srbcheema1 replied:
Hmm thats right, I will use the same function.

srbcheema1 commented on docker/Dockerfile:
```diff
@@ -5,6 +5,7 @@ g++ \
 make \
 cmake \
 libbz2-dev \
+libboost-all-dev \
```
Oh thats my fault, I just added that for my local ease. Just forgot to remove that. :sweat_smile: 
Yes I use docker build.

srbcheema1 commented on inc/vcf/validator.hpp:
```diff
@@ -185,12 +189,16 @@ namespace ebi
                            std::vector<std::unique_ptr<ebi::vcf::ReportWriter>> &outputs);
 
-    bool is_compressed_file(const std::string &source,
+    std::string is_compressed_file(const std::string &source,
```
I got that. I will take care of naming in future.

srbcheema1 commented on src/vcf/validator.cpp:
```diff
@@ -135,12 +136,26 @@ namespace ebi
                            ValidationLevel validationLevel,
                            std::vector<std::unique_ptr<ebi::vcf::ReportWriter>> &outputs)
     {
+        std::string file_ext = ebi::vcf::is_compressed_extension(sourceName);
```
@jmmut first argument should not be const as there . As the line
        `in.push(input);`
https://github.com/srbcheema1/vcf-validator/blob/91a05b82442a594fa096233f045618a863a93f41/src/vcf/validator.cpp#L226 
will cause error as there will be no matching function for the call `push(const istream&)`

jmmut commented on inc/vcf/validator.hpp:
```diff
@@ -185,12 +189,16 @@ namespace ebi
 
-    bool is_compressed_file(const std::string &source,
-                            const std::vector<char> &line);
+    std::string get_compressed_file(const std::string &source,
```
alright, sorry to ask for another renaming, but I don't think the current names are good enough. what about `get_compression`, `get_compression_from_extension` and `get_compression_from_magic_num`?

jmmut commented on src/vcf/validator.cpp:
```diff
@@ -181,7 +179,24 @@ namespace ebi
+
+        if (!readable_extentions.count(file_ext)) {
+            throw std::invalid_argument{"Input file should not be compressed"};
+        }
```
I can see that this function will need to be modified to ask for the first line and check for the magic num as well.

It's ok until you add that.

jmmut commented on src/vcf/validator.cpp:
```diff
@@ -135,12 +135,19 @@ namespace ebi
-        if (ebi::vcf::is_compressed_file(sourceName, line)) {
+        ebi::util::readline(uncompressed_input, line);
+
+        std::string compression_type = ebi::vcf::get_compressed_magic_num(line);
```
This 4 lines are checking if the file is compressed twice. it can go in another function

srbcheema1 commented:
@jmmut now the vcf-validator is capable to read bzip and gzip compressions directly.
please have a look on the workflow.

Now it works like this.


```
$ cat compressed.vcf.gz | ./vcf_validator
[info] Reading from standard input...
[warning] Detected .gz compression
[info] Report written to : stdin.errors_summary.1521373523880.txt
[info] According to the VCF specification, the input file is valid
```


```
$ ./vcf_validator < compressed.vcf.bz2
[info] Reading from standard input...
[warning] Detected .bz2 compression
[info] Report written to : stdin.errors_summary.1521373404396.txt
[info] According to the VCF specification, the input file is valid
```


```
$ ./vcf_validator -i compressed.vcf.gz
[info] Reading from input file...
[warning] Detected .gz compression
[info] Report written to : compressed.vcf.gz.errors_summary.1521383548014.txt
[info] According to the VCF specification, the input file is valid
```

srbcheema1 commented on src/vcf/validator.cpp:
```diff
@@ -150,38 +164,76 @@ namespace ebi
+        }
+
+        for (int j = 0; j < (int)container.size(); ++j) {
+            stream.unget();
```
@jmmut I have added a check to throw if stream fails.
Reason why I am not preferring reading first line is that in case of a compressed stream the size of `first_line`  can be very large. I think function `read_line` will almost read the whole stream in one go, if the stream is compressed.
The `first_line` may not be used directly as it may be compressed(in case of compressed stream). We will again have to send it for filtering. That will be equivalent to the approach I proposed earlier in beginning of this PR.

jmmut replied:
not exactly.

- In case it is not compressed, you will read either the first 4 bytes of the "##fileformat" line or until "\n" if the line is invalid and shorter than 4 bytes, and that string should be used as-is. In fact, if you find a "\n" in those 4 bytes the `detect_version` function should throw an exception later.
- In case it is compressed, you will read the 4 bytes and only need the subsequent bytes to send them through the filtering_stream. In this case, we should not read until "\n" is found because if it's compressed, the newlines won't appear as such.

jmmut commented on src/vcf/validator.cpp:
```diff
@@ -150,38 +164,79 @@ namespace ebi
+        for (int j = 0; j < (int)container.size(); ++j) {
+            stream.unget();
+            if (stream.fail() || !stream) {
+                throw std::runtime_error("Stream failed");
```
Add more context to this message. Imagine when will this message show up.

hypothetical situation in 12 months in the future:
> Ok I refactored the main.cpp and the validator.cpp a bit, but all tests should work. Oh they failed, and any validation is only showing:
```
$ cat my.vcf.gz |  ./vcf_validator 
Stream failed
$ 
```

srbcheema1 commented:
@jmmut I have replaced the `unget()` method with more reliable `putback()` method.

boost provides [putback testcases](https://github.com/boostorg/iostreams/blob/develop/test/putback_test.hpp) to ensure validity of multiple putbacks possible.
Still for our safety we have inserted a check condition to check the failbit of stream.

I have also modified the message and made it more descriptive.
Please have a look.

jmmut commented on src/vcf/validator.cpp:
```diff
@@ -198,10 +198,11 @@ namespace ebi
+            stream.putback(container[j]);
             if (stream.fail() || !stream) {
-                throw std::runtime_error("Stream failed");
+                throw std::runtime_error("Stream failed while putting the magic_numbers back into stream \
```
this message will show as 
```
Stream failed while putting the magic_numbers back into stream
                        while reading magic numbers
```
keep only the first line, please

srbcheema1 commented:
@jmmut Done!

I want to discuss something regarding the log message displayed while catching `std::runtime_error`.

https://github.com/EBIvariation/vcf-validator/blob/6e2a7145d367d4a5705f31d8d0d6ed9380b0028e/src/validator_main.cpp#L201

I noticed that the runtime error always show `file is not valid`. Is it always true that file will be invalid if we catch a runtime error? Maybe a runtime error occurs on a valid file. Like in case of this line 
https://github.com/EBIvariation/vcf-validator/blob/6e2a7145d367d4a5705f31d8d0d6ed9380b0028e/src/validator_main.cpp#L135
Should we display `ex.what()` only?

jmmut commented:
Well, the issue about the message is true, It should probably be something along the lines: "The validation could not be completed: " and then the inner message.

srbcheema1 commented on -.s:
```diff
@@ -0,0 +1 @@
+	.file	""
```
Sorry that was my mistake. It went unnoticed 😂

srbcheema1 commented on src/vcf/validator.cpp:
```diff
@@ -135,12 +142,19 @@ namespace ebi
                            std::vector<std::unique_ptr<ebi::vcf::ReportWriter>> &outputs)
     {
+        boost::iostreams::filtering_istream uncompressed_input;
+        boost::iostreams::filtering_istream input_stream(input);
```
filtering_stream::push doesn't make copy of it. filtering_stream is not copy-constructible. So if we declare `input_stream` in function`create_uncompressed_stream` it will be destroyed with the function return and will cause segmentation fault.

srbcheema1 commented:
@jmmut  I have made the requested changes and addressed the comments.

jmmut commented on src/vcf/validator.cpp:
```diff
@@ -150,54 +162,113 @@ namespace ebi
+        std::string file_ext = get_compression(sourceName, line);
+        check_readability_of_file(file_ext);
+
+        if(file_ext == BZ2) {
```
I saw boost provides also `boost::iostreams::zlib_decompressor`, maybe we can add `.zip` as well to readable files?

srbcheema1 commented:
@jmmut I have added zlib_decompressor to support reading .z file compression.

Anishka0107 commented:
I will soon add the assembly_checker tool, and I think it would be great if this code could decompress the files (vcf, fasta) we get there.

Maybe the detection and decompression code could be in a separate class, so that other tools can easily reuse it?

WDYT?

jmmut commented:
@Anishka0107 unless it's strictly necessary, try to keep it independent, work from the develop branch. Don't worry about including the decompression functionality. We'll get there at some point.

srbcheema1 commented:
Should I make more changes in this PR?

cyenyxe commented on src/vcf/validator.cpp:
```diff
@@ -150,54 +162,116 @@ namespace ebi
+        for (int j = (int)container.size()-1; j >= 0; --j) {
+            stream.putback(container[j]);
+            if (stream.fail()) {
+                throw std::runtime_error("Stream failed while putting the magic_numbers back into stream");
```
`magic numbers` instead of `magic_numbers`

srbcheema1 commented:
@cyenyxe  I have done the requested changes.

eroller commented:
> @jmmut now the vcf-validator is capable to read bzip and gzip compressions directly.
> Please have a look.
> 
> Now it works like this.
> 
> ```
> $ cat compressed.vcf.gz | ./vcf_validator
> [info] Reading from standard input...
> [warning] Detected .gz compression
> [info] Report written to : stdin.errors_summary.1521373523880.txt
> [info] According to the VCF specification, the input file is valid
> ```
> 
> ```
> $ ./vcf_validator < compressed.vcf.bz2
> [info] Reading from standard input...
> [warning] Detected .bz2 compression
> [info] Report written to : stdin.errors_summary.1521373404396.txt
> [info] According to the VCF specification, the input file is valid
> ```
> 
> ```
> $ ./vcf_validator -i compressed.vcf.gz
> [info] Reading from input file...
> [warning] Detected .gz compression
> [info] Report written to : compressed.vcf.gz.errors_summary.1521383548014.txt
> [info] According to the VCF specification, the input file is valid
> ```

why is there still a warning about .gz compression?

jmmut commented:
@eroller you are right that it doesn't make sense anymore to flag that as a warning. It should probably an info or even a debug log. I took note to change this but I can't guarantee we can work on this soon.

srbcheema1 commented:
@jmmut @eroller @cyenyxe I am not sure if it is fully solved, I think its not readable on windows platform fully.
https://github.com/EBIvariation/vcf-validator/blob/0b09fa9bbe8392cab6aa5da1d12d9670ff50c274/test/vcf/compressed_file_test.cpp#L72

jmmut commented:
@srbcheema1 in windows it has to complain because it's not supported, but in linux/mac it doesn't make sense to warn about it, it can be a debug there.

srbcheema1 commented:
I think better to make the message to `debug` in mac/linux and `warning` in windows.

jmmut commented:
I'm not 100% sure that the code is doing this, but if you provide a compressed VCF to the windows program, it should not run. It should complain and stop. It doesn't make sense to just issue a warning and continue processing it, because we know compression doesn't work in windows.
